api/src/main/java/com/microsoft/gctoolkit/GCToolKit.java

```diff
@@ -127,13 +88,136 @@ public void loadAggregationsFromServiceLoader() {
+                        .stream()
+                        .map(ServiceLoader.Provider::get)
+                        .findFirst()
+                        .orElseThrow(() -> new ServiceConfigurationError("Internal Error - No suitable JVMEventBus implementation found")));
+            } else {
+                try {
+                    Class clazz = forName("com.microsoft.gctoolkit.vertx.VertxJVMEventChannel", true, Thread.currentThread().getContextClassLoader());
+                    loadJVMEventChannel((JVMEventChannel) clazz.getConstructors()[0].newInstance());
+                } catch (Exception e) {
+                    throw new ServiceConfigurationError("Unable to find a suitable provider to create a JVMEventChannel");
+                }
+            }
+        }
+    }
+
+    public void loadDataSourceParser(DataSourceParser dataSourceParser) {
+        registeredDataSourceParsers.add(dataSourceParser);
+    }
+
+    private void loadDataSourceParsers(Diary diary) {
+        loadDataSourceChannel();
+        loadJVMEventChannel();
+        List<DataSourceParser> dataSourceParsers;
+        if (registeredDataSourceParsers.isEmpty()) {
+            dataSourceParsers = ServiceLoader.load(DataSourceParser.class)
+                    .stream()
+                    .map(ServiceLoader.Provider::get)
+                    .filter(consumer -> consumer.accepts(diary))
+                    .collect(Collectors.toList());
+        } else {
+            dataSourceParsers = new ArrayList<>();
+            dataSourceParsers.addAll(registeredDataSourceParsers);
+        }
+
+
+        if (dataSourceParsers.isEmpty()) {
+            String[] parsers = {
+                    "com.microsoft.gctoolkit.parser.CMSTenuredPoolParser",
+                    "com.microsoft.gctoolkit.parser.GenerationalHeapParser",
+                    "com.microsoft.gctoolkit.parser.JVMEventParser",
+                    "com.microsoft.gctoolkit.parser.PreUnifiedG1GCParser",
+                    "com.microsoft.gctoolkit.parser.ShenandoahParser",
+                    "com.microsoft.gctoolkit.parser.SurvivorMemoryPoolParser",
+                    "com.microsoft.gctoolkit.parser.UnifiedG1GCParser",
+                    "com.microsoft.gctoolkit.parser.UnifiedGenerationalParser",
+                    "com.microsoft.gctoolkit.parser.UnifiedJVMEventParser",
+                    "com.microsoft.gctoolkit.parser.UnifiedSurvivorMemoryPoolParser",
+                    "com.microsoft.gctoolkit.parser.ZGCParser"
+            };
+            try {
+                dataSourceParsers = Arrays.stream(parsers)
+                        .map(parserName -> {
+                            try {
+                                return forName(parserName, true, Thread.currentThread().getContextClassLoader());
+                            } catch (ClassNotFoundException e) {
+                                throw new ServiceConfigurationError("Unable to find a suitable provider to create a DataSourceParser");
```

Comment:
I suggest replacing the whole try-catch block starting on line 189 with this: 
```java
dataSourceParsers = Arrays.stream(parsers)
                        .map(parserName -> {
                            try {
                                Class<?> clazz = forName(parserName, true, Thread.currentThread().getContextClassLoader());
                                return Optional.of(clazz.getConstructors()[0].newInstance());
                            } catch (ClassNotFoundException 
                                     | InstantiationException 
                                     | IllegalAccessException 
                                     | InvocationTargetException e) {
                                return Optional.empty();
                            }
                        })
                        .filter(Optional::isPresent)
                        .map(optional -> (DataSourceParser) optional.get())
                        .filter(consumer -> consumer.accepts(diary))
                        .collect(Collectors.toList());
                if (dataSourceParsers.isEmpty()) {
                    throw new ServiceConfigurationError("Unable to find a suitable provider to create a DataSourceParser");
                } 
```
which would loop through the whole parsers array instead of bailing on the first exception.